### PR TITLE
Switched sync.{cpp,h} to std threading primitives.

### PR DESCRIPTION
**Pull request by tjps**
Replaced boost threading primitives with the std equivalents.

### CODE REVIEW

**Comment by sipa** [comment]
This uses `thread_local`, a feature that we chose to not use before, though I don't remember why. Perhaps because we needed compatibility with GCC 4.7? Is that still the case, @laanwj @theuni

**Comment by laanwj** [comment]
> This uses thread_local, a feature that we chose to not use before, though I don't remember why. 

Given that it's only used when lock debugging is used I'd say it's acceptable.

Also FWIW I tried to build bitcoin on g++ 4.7 a while ago and it was a pain; for example the `AX_CXX_COMPILE_STDCXX` fails because the ` __cplusplus` macro is not set correctly, and that's only the beginning. As it's already effectively broken and no one else seems to even notice it, I wouldn't mind bumping the requirement to g++ 4.8 wholesale.

File: src/sync.cpp [hunk 1]
```diff
@@ -4,13 +4,12 @@
 
 #include <sync.h>
 
+#include <set>
```
Comment (promag) [inline src/sync.cpp]: Why?

**Review by promag — COMMENTED** [review]
Concept ACK.

File: src/sync.cpp (hunk 1 above)
Comment (tjps) [reply src/sync.cpp]: Amazingly enough the #include tree in sync.cpp did not pull in std::set already.

**Comment by practicalswift** [comment]
Concept ACK

**Comment by tjps** [comment]
Since #11732 was resolved in favor of bumping GCC requirement to 4.8, looks like this PR is good to go?

**Comment by sipa** [comment]
utACK bba9bd0d9dd06f13a6b0c89181864453cab5c858

**Review by theuni — COMMENTED** [review]
thread_local was also off-limits due to missing support in osx's clang. It's been added as of XCode 8 on September 13, 2016. I'm uneasy with that as a minimum requirement.

Additionally, last time I checked this, the use of thread_local causes new symbols to be pulled in from glibc on Linux, meaning that we would lose back-compat with older versions. I'd have to re-test to see where the break lies.

That said, I think it's ok to lessen the requirement for the DEBUG_LOCKCONTENTION case, as long as there's a comment there mentioning that it's problematic for other code.

As osx is probably not the only platform where this will cause headaches, I'd also prefer to add an autoconf link test for thread_local.

**Comment by TheBlueMatt** [comment]
utACK bba9bd0d9dd06f13a6b0c89181864453cab5c858 modulo thread_local concerns. I also think its probably OK in DEBUG_LOCKCONTENTION only even if its not on all supported platforms.

**Comment by theuni** [comment]
Pushed up a commit on top of this PR that adds the configure check for thread_local: https://github.com/theuni/bitcoin/commit/d6df1162d3cba1aabae4f5ae61da6a81881ea118 . I've verified that it works as intended for me in Linux, and breaks as intended on osx.

@tjps assuming that works for you, mind cherry-picking it here?

**Comment by tjps** [comment]
That looks great to me, will cherry-pick it onto this PR.

**Review by theuni — APPROVED** [review]
Thanks! utACK f7f7e2cd340c088e82d09090eb275b98b34a9812.

**Pull request merged**